### DISCUSSION TITLE
zlibrary plugin.

### DISCUSSION THREAD

**Discussion by greatyingzi** [koreader/koreader#10242] [Ideas]
If someone make a zlibrary plug-in for browsing and downloading books, is it allowed to merge ? @NiLuJe

**Comment by Frenzie** [comment]
No, that's not something we can do.

**Reply by greatyingzi** [reply]
> No, that's not something we can do.

@Frenzie  we can.just like https://github.com/baroxyton/zlibrary-CLI,or 
https://github.com/sertraline/zlibrary.  we can parser html to get infos(cover image, title , author, download link) and show theme by lua

**Reply by NiLuJe** [reply]
Nope.

**Reply by NiLuJe** [reply]
(I mean, you're free to do whatever you like, but it will neither hit the main nor the contrib repo, or even the wiki).

**Reply by poire-z** [reply]
Your question wasn't about whether it's doable technically (I'm quite sure _someone can make a zlibrary plug-in for browsing and downloading books_), but about whether, once it is implemented, we would welcome the plugin in the project and ship it with KOReader.
And the answer is: no.

Because of the possibly illegal/moral status of zlibrary for some parties/countries/users - we'd rather not have to think about it and be involved and take sides and spend time arguing about legal questions (we're developpers), and compromise this project. It's not essential to KOReader itself.
You're totally free, outside of the koreader repositories, to write such a plugin and share it with the users interested - but it's you and they who will ask and answer the above questions and make decisions - we don't wanna know :)

**Reply by greatyingzi** [reply]
thanks.

**Reply by greatyingzi** [reply]
Can we make a plugin manager that allows users to add third-party plugins via network connection?Let the plug-in and the main program decoupling, separate maintenance. @poire-z

**Reply by poire-z** [reply]
I think we could, and we have discussed that in the past (but can't find the discussion/issue) - but it's also not high priority: we ship many plugins, others are available at https://github.com/koreader/contrib/, and it's a one time operation: download and unzip on your device, not worth the time writing something to do that :)
But it would help with keeping such plugins update - but then we'd need to manage versions (ie. some plugin version may only work with a newer koreader version).

Anyway, we'd have to manage somewhere a list of the plugins available to that plugin manager - and for the same reasons I explained above, we probably would not even include a zlibary plugin in that list :)

**Reply by greatyingzi** [reply]
There is indeed such a problem, but I don't think it's unsolvable. The official only needs to ensure that its own plugins and the non-official plugins included (legitimate and active, of course, with some other conditions) are updated with the main version of koreader. As for third-party non-included plugins, it should be the responsibility of their maintainers. I think this solution is quite feasible, at least various Linux distributions currently adopt similar practices.

**Reply by Frenzie** [reply]
Yes, we understand the concept. You're extremely close to describing the situation as is.

**Reply by greatyingzi** [reply]
Do we have a plan to create such a plugin manager? I'm really looking forward to it!

**Reply by NiLuJe** [reply]
IIRC, @pazos has been working on and off on the idea for a while.

**Reply by pazos** [reply]
lol, that's true! still WIP, NSFW et al.

I managed to build a huge Frankenstein-like monster of spec, with containers,
manifests, packages, bundles...

I've got sidetracked by trying to add signatures to packages so they can be verified. Of course this had to happen outside of the container/manifest so I've padded a little fixed size header that can be fseek'd out of the equation, so the end user can lurk on the (zip) container and read the manifest within it.

This looks cool inside the app but makes binaries a bit obscure outside KO.

**Reply by pazos** [reply]
A `container` is, until now, a zip file.
A manifest is a plain text file named `manifest` on the root of the zip file. It describes the container contents.

A `package` is every container that contains a valid manifest. 

A `bundle` is a kind of package that contains installable content.

Bundles might be "offline" (ie: all the code is inside the container) or "online" (manifest has links to git repos that can be pulled from the internet) or a mix of them.

I didn't integrate libgit yet, didn't wrote the UI bits to preview packages or use them. It's almost in the design phase but thanks to https://github.com/koreader/koreader/pull/9615 I can state that offline bundles work (kind of)

**Reply by pazos** [reply]
Hopefully I will get a RFC prototype after my summer holidays :)

**Reply by UntriexTv** [reply]
Is there any update regarding this?

**Reply by pazos** [reply]
Not really :/

It was a fun ride but nothing maintenable emerged.

On the bright side: plugins maintained elsewhere seem to be widespread if they serve a purpose for people out there. Cases in point are Rakuyomi, Assistant or Anki tools. These plugins have a long life on its own.
